gateway/downlink.go

```diff
@@ -0,0 +1,212 @@
+		return fmt.Errorf("error sending downlink: %w", ctx.Err())
+	}
+
+	errCode := int(C.send(&txPkt))
+	if errCode != 0 {
+		return errors.New("failed to send downlink packet")
+	}
+	// wait for packet to finish sending.
+	var status C.uint8_t
+	for {
+		C.lgw_status(txPkt.rf_chain, 1, &status)
+		if err := ctx.Err(); err != nil {
+			return fmt.Errorf("error sending downlink: %w", ctx.Err())
+		}
+		// status of 2 indicates send was successful
+		if int(status) == 2 {
+			break
+		}
+		time.Sleep(2 * time.Millisecond)
+	}
+
+	g.logger.Infof("sent the downlink packet")
+
+	return nil
+}
+
+// According to lorawan docs, downlinks have a +/- 20 us error window, so regular sleep would not be accurate enough.
+func accurateSleep(ctx context.Context, duration time.Duration) bool {
+	// If we use utils.SelectContextOrWait(), we will wake up sometime after when we're supposed
+	// to, which can be hundreds of microseconds later (because the process scheduler in the OS only
+	// schedules things every millisecond or two). For use cases like a web server responding to a
+	// query, that's fine. but when outputting a PWM signal, hundreds of microseconds can be a big
+	// deal. To avoid this, we sleep for less time than we're supposed to, and then busy-wait until
+	// the right time. Inspiration for this approach was taken from
+	// https://blog.bearcats.nl/accurate-sleep-function/
+	// On a raspberry pi 4, naively calling utils.SelectContextOrWait tended to have an error of
+	// about 140-300 microseconds, while this version had an error of 0.3-0.6 microseconds.
+	startTime := time.Now()
+	maxBusyWaitTime := 1500 * time.Microsecond
+	if duration > maxBusyWaitTime {
+		shorterDuration := duration - maxBusyWaitTime
+		if !utils.SelectContextOrWait(ctx, shorterDuration) {
+			return false
+		}
+	}
+
+	for time.Since(startTime) < duration {
+		if err := ctx.Err(); err != nil {
+			return false
+		}
+		// Otherwise, busy-wait some more
+	}
+	return true
+}
+
+// Downlink payload structure
+// | MHDR | DEV ADDR | FCTRL | FCNTDOWN |  FOPTS (optional)  |  FPORT | encrypted frame payload  |  MIC |
+// | 1 B  |   4 B    |  1 B  |    2 B   |       variable     |   1 B  |      variable            | 4 B  |
+func (g *gateway) createDownlink(device *node.Node, framePayload []byte) ([]byte, error) {
+	payload := make([]byte, 0)
+
+	// Mhdr unconfirmed data down
+	payload = append(payload, 0x60)
+
+	devAddrLE := reverseByteArray(device.Addr)
+
+	payload = append(payload, devAddrLE...)
+
+	// 3. FCtrl: ADR (0), RFU (0), ACK (0), FPending (0), FOptsLen (0)
+	payload = append(payload, 0x00)
+
+	fCntBytes := make([]byte, 2)
+	binary.LittleEndian.PutUint16(fCntBytes, uint16(device.FCntDown)+1)
+	payload = append(payload, fCntBytes...)
+
+	// fopts are used for the MAC commands
+	// fopts := []byte{
+	// 	0b00111001, // data rate and tx power
+	// 	0xFF,
+	// 	0x00,
+	// 	0x01,
+	// }
+
+	// fopts := []byte{0x3A, 0xFF, 0x00, 0x01}
+
+	// payload = append(payload, fopts...)
```

Comment:
[note to self] we wanted to keep these comments for future testing when we support fopts